File: frontend/src/modules/documents/services/documentService.ts
import api from '@/core/api/client';
import type {
  Document,
  CreateDocumentRequest,
  UpdateDocumentRequest,
  BackupFile,
  SyncIssue,
  ReloadResponse,
} from '@/types';

export const DocumentService = {
  getAll: (offset = 0, limit = 15, sort_by?: string, sort_desc?: boolean) => 
    api.get<Document[]>('/api/documents', { params: { offset, limit, sort_by, sort_desc } }),

  search: (params: {
    uuid?: string;
    code?: string;
    folder_name?: string;
    field_key?: string;
    date_from?: string;
    date_to?: string;
    [key: string]: any;
    sort_by?: string;
    sort_desc?: boolean;
    offset?: number;
    limit?: number;
  }) => {
    const cleanParams: any = {};
    Object.entries(params).forEach(([key, value]) => {
      // Allow field_* parameters even if empty, but filter out other empty values
      if (value !== undefined && value !== null) {
        if (key.startsWith('field_') || value !== '') {
          cleanParams[key] = value;
        }
      }
    });
    return api.get<Document[]>('/api/documents/search', { params: cleanParams });
  },

  getOne: (uuid: string) => 
    api.get<Document>(`/api/documents/${uuid}`),

  getById: (uuid: string) =>
    api.get<Document>(`/api/documents/${uuid}`),

  getByCode: (code: string) =>
    api.get<Document>(`/api/documents/code/${code}`),

  // Only exposes admin methods if generic type allows, but API enforces security
  create: (data: CreateDocumentRequest) => 
    api.post<Document>('/api/documents', data),

  update: (uuid: string, data: UpdateDocumentRequest) =>
    api.put<Document>(`/api/documents/${uuid}`, data),

  remove: (uuid: string) =>
    api.delete<void>(`/api/documents/${uuid}`),
    
  uploadFile: (uuid: string, file: File) => {
    const formData = new FormData();
    formData.append('file', file);
    return api.post(`/api/documents/${uuid}/files`, formData, {
      headers: { 'Content-Type': 'multipart/form-data' }
    });
  },

  /**
   * Downloads a file with proper authentication.
   * Returns the file as a blob so it can be downloaded with auth headers.
   */
  downloadFile: async (uuid: string, fileName: string) => {
    const response = await api.get(`/api/documents/${uuid}/files/${fileName}`, {
      responseType: 'blob'
    });
    return response.data;
  },

  /**
   * Updates the description and notes for a specific file entry in files.json.
   * Matches: PUT /documents/{uuid}/files/{fileName}
   */
  updateFileMetadata: (uuid: string, fileName: string, data: { description: string; note: string }) =>
    api.put(`/api/documents/${uuid}/files/${fileName}`, data),

  /**
   * Deletes both the physical file and its entry in files.json.
   * Matches: DELETE /documents/{uuid}/files/{fileName}
   */
  deleteFile: (uuid: string, fileName: string) =>
    api.delete(`/api/documents/${uuid}/files/${fileName}`),

  /**
   * Updates the file contents directly.
   * Matches: PATCH /documents/{uuid}/files/{fileName}/contents
   */
  updateFileContents: (uuid: string, fileName: string, data: { contents: string }) =>
    api.patch(`/api/documents/${uuid}/files/${fileName}/contents`, data),

  /**
   * Renames a file.
   * Matches: PATCH /documents/{uuid}/files/{fileName}/rename
   */
  renameFile: (uuid: string, fileName: string, data: { new_name: string }) =>
    api.patch(`/api/documents/${uuid}/files/${fileName}/rename`, data),

  /**
   * Lists all sync conflicts (admin only).
   * Matches: GET /documents/conflicts
   */
  listConflicts: () =>
    api.get<SyncIssue[]>('/api/documents/conflicts'),

  /**
   * Reloads all documents (admin only).
   * Matches: POST /documents/reload
   */
  reload: () =>
    api.post<ReloadResponse>('/api/documents/reload'),

  /**
   * Reloads a specific folder/document (admin only).
   * Matches: POST /documents/reload/{folderName}
   */
  reloadDocument: (folderName: string) =>
    api.post<ReloadResponse>(`/api/documents/reload/${folderName}`),

  /**
   * Lists available backup files.  
   * Matches: GET /api/backup
   */
  listBackups: () =>
    api.get<{ backups: BackupFile[] }>('/api/documents/backup'),

  /**
   * Downloads a specific backup file by name.
   * Matches: GET /api/backup/download/{fileName}
   */
  downloadBackup: async (fileName: string) => {
    const response = await api.get(`/api/documents/backup/download/${fileName}`, {
      responseType: 'blob'
    });
    return response.data;
  },

  /**
   * Triggers the creation of backup.
   * Returns a job_id for progress tracking.
   */
  createBackup: async () => {
    const response = await api.post<{ job_id: string; message: string }>('/api/documents/backup');
    return response.data;
  },

  /**
   * Restores a backup file that is already on the server.
   * Returns a JobID for progress tracking.
   */
  restoreBackup: (fileName: string, mode: 'merge' | 'overwrite' = 'merge') => {
    return api.post<{ 
      job_id: string; 
      message: string; 
    }>(`/api/documents/backup/restore?file_name=${fileName}&mode=${mode}`);
  },

  /**
   * Polls the status of any background job (backup or restore).
   * Matches: GET /api/jobs/{jobID}
   */
  getJobStatus: (jobId: string) => {
    return api.get<{
      type: 'backup' | 'restore';
      progress: number;
      status: 'processing' | 'completed' | 'failed';
      error?: string;
    }>(`/api/jobs/${jobId}`); 
  },
};

File: server/core/document/document_service.go
package document

import (
	"context"
	"errors"
	"fmt"
	"io"
)

type DocumentService struct {
	repo DocumentRepository
}

func NewDocumentService(repo DocumentRepository) *DocumentService {
	return &DocumentService{repo: repo}
}

func (s *DocumentService) Create(ctx context.Context, doc Document) (*Document, error) {
	if err := validateDocumentTitle(doc.Title); err != nil {
		return nil, err
	}

	return s.repo.Create(ctx, &doc)
}

func (s *DocumentService) GetByUUID(ctx context.Context, uuid string) (*Document, error) {
	if uuid == "" {
		return nil, errors.New("uuid is required")
	}

	doc, err := s.repo.GetByUUID(ctx, uuid)
	if err != nil {
		return nil, fmt.Errorf("fetching document %s: %w", uuid, err)
	}

	return doc, nil
}

func (s *DocumentService) GetByCode(ctx context.Context, code string) (*Document, error) {
	if code == "" {
		return nil, errors.New("code is required")
	}

	return s.repo.GetByCode(ctx, code)
}

func (s *DocumentService) Update(ctx context.Context, uuid string, doc Document) (*Document, error) {
	if uuid == "" {
		return nil, errors.New("uuid is required")
	}
	if err := validateDocumentTitle(doc.Title); err != nil {
		return nil, err
	}

	return s.repo.Update(ctx, uuid, &doc)
}

func (s *DocumentService) Delete(ctx context.Context, uuid string) error {
	if uuid == "" {
		return errors.New("uuid is required")
	}

	return s.repo.Delete(ctx, uuid)
}

func (s *DocumentService) List(ctx context.Context, offset int, limit int, sortBy string, sortDesc bool) ([]*Document, error) {
	return s.repo.List(ctx, offset, limit, sortBy, sortDesc)
}

func (s *DocumentService) UploadFile(ctx context.Context, uuid string, fileName string, content io.Reader) error {
	if uuid == "" {
		return errors.New("uuid is required")
	}
	if fileName == "" {
		return errors.New("file name is required")
	}
	if content == nil {
		return errors.New("file content is required")
	}

	// Delegate to repository which handles path construction, security, and file I/O
	return s.repo.UploadFile(ctx, uuid, fileName, content)
}

func (s *DocumentService) DownloadFile(ctx context.Context, uuid string, fileName string) (io.ReadCloser, error) {
	if uuid == "" || fileName == "" {
		return nil, errors.New("uuid and fileName are required")
	}
	return s.repo.DownloadFile(ctx, uuid, fileName)
}

func (s *DocumentService) UpdateFileMetadata(ctx context.Context, uuid string, fileName string, updates FileMetadataUpdate) error {
	if uuid == "" || fileName == "" {
		return errors.New("uuid and fileName are required")
	}
	return s.repo.UpdateFileMetadata(ctx, uuid, fileName, updates)
}

func (s *DocumentService) DeleteFile(ctx context.Context, uuid string, fileName string) error {
	if uuid == "" {
		return errors.New("uuid is required")
	}
	if fileName == "" {
		return errors.New("file name is required")
	}

	return s.repo.DeleteFile(ctx, uuid, fileName)
}

func (s *DocumentService) UpdateFileContents(ctx context.Context, uuid string, fileName string, content io.Reader) error {
	if uuid == "" {
		return errors.New("uuid is required")
	}
	if fileName == "" {
		return errors.New("file name is required")
	}
	if content == nil {
		return errors.New("file content is required")
	}

	return s.repo.UpdateFileContents(ctx, uuid, fileName, content)
}

func (s *DocumentService) RenameFile(ctx context.Context, uuid string, oldName string, newName string) error {
	if uuid == "" {
		return errors.New("uuid is required")
	}
	if oldName == "" || newName == "" {
		return errors.New("old and new file names are required")
	}

	return s.repo.RenameFile(ctx, uuid, oldName, newName)
}

func (s *DocumentService) GetConflicts(ctx context.Context) ([]SyncIssue, error) {
	return s.repo.GetConflicts(ctx)
}

func (s *DocumentService) ReloadCache(ctx context.Context) ([]SyncIssue, error) {
	return s.repo.ReloadCache(ctx)
}

func (s *DocumentService) ReloadCacheForFolder(ctx context.Context, folderName string) ([]SyncIssue, error) {
	return s.repo.ReloadCacheForFolder(ctx, folderName)
}

func (s *DocumentService) Search(ctx context.Context, criteria SearchCriteria) ([]*Document, error) {
	return s.repo.Search(ctx, criteria)
}

func (s *DocumentService) GetBackupPath() string {
	return s.repo.GetBackupPath()
}

func (s *DocumentService) ListBackups(ctx context.Context) ([]BackupFile, error) {
	return s.repo.ListBackups(ctx)
}

func (s *DocumentService) CreateBackup(ctx context.Context) (string, error) {
	return s.repo.CreateBackup(ctx)
}

func (s *DocumentService) RestoreFromLocalPath(ctx context.Context, fileName string, overwrite bool, onProgress func(float64)) error {
	return s.repo.RestoreFromLocalPath(ctx, fileName, overwrite, onProgress)
}

func validateDocumentTitle(title string) error {
	if title == "" {
		return errors.New("document title is required")
	}

	return nil
}

func getFileExtension(fileName string) string {
	for i := len(fileName) - 1; i >= 0; i-- {
		if fileName[i] == '.' {
			return fileName[i:]
		}
	}
	return ""
}
